builtin/update-index.c

```diff
@@ -233,6 +233,8 @@ static int mark_ce_flags(const char *path, int flag, int mark)
 	int namelen = strlen(path);
 	int pos = cache_name_pos(path, namelen);
 	if (0 <= pos) {
```

Comment:
Commit message thoughts:

Topic should focus on the builtin first:

```
update-index: integrate with sparse index
```

See [`SubmittingPatches`](https://github.com/git/git/blob/master/Documentation/SubmittingPatches#L103) and specifically "imperative mood" to avoid describing "this change" when possible.

```
Allow 'git update-index' to operate directly on a sparse index. Additional
checks are required to ensure we do not match the user-specified path
to a sparse directory entry...
```

This raises an interesting question: we couldn't update a directory before because the index did not contain entries for directories. However, we now have directories, but they are always sparse. If `git update-index` allowed specifying an OID instead of caring about the working tree, then I could see an interesting case for allowing a user to modify the sparse directory entry to another tree (with an appropriate `--sparse` option or something). As it is, I don't think that is something we should worry about.